db/common/src/main/java/org/commonjava/indy/db/common/AbstractStoreDataManager.java

```diff
@@ -422,4 +442,73 @@ public void disableNotValidStore(ArtifactStore store,ArtifactStoreValidateData v
         return streamArtifactStoreKeys().filter( key -> key.getPackageType().equals( pkg ) && key.getType() == type )
                                         .collect( Collectors.toSet() );
     }
+
+    public Set<Group> affectedBy( final Collection<StoreKey> keys )
+            throws IndyDataException
+    {
+        return affectedByFromStores( keys );
+    }
+
+    protected Set<Group> affectedByFromStores( final Collection<StoreKey> keys )
+            throws IndyDataException
+    {
+        Logger logger = LoggerFactory.getLogger( getClass() );
+        logger.debug( "Getting groups affected by: {}", keys );
+
+        List<StoreKey> toProcess = new ArrayList<>( new HashSet<>( keys ) );
+
+        Set<Group> groups = new HashSet<>();
+        if ( toProcess.isEmpty() )
+        {
+            return groups;
+        }
+
+        Set<StoreKey> processed = new HashSet<>();
+        final String packageType = toProcess.get( 0 ).getPackageType();
+
+        Set<ArtifactStore> all = this.getAllArtifactStores().stream().filter( st -> {
+            if ( packageType != null && !st.getPackageType().equals( packageType ) )
+            {
+                return false;
+            }
+
+            if ( st.getType() != StoreType.group )
+            {
+                return false;
+            }
+
+            return true;
+        } ).collect( Collectors.toSet() );
```

Comment:
As we've involved that storesByPkg cache, I think this part of code should be refactored to use it like below:

        Set<StoreKey> allKeys = this.getStoreKeysByPkgAndType( toProcess.get( 0 ).getPackageType(), group );
        logger.debug( "There are {} groups need to loop checking for affected by", all.size() );

        Set<ArtifactStore> all = allKeys.stream().map( k-> {
            try
            {
                return this.getArtifactStore( k );
            }
            catch ( IndyDataException e )
            {
                logger.error( "Error to get store {}", k );
                return null;
            }
        } ).filter( Objects::nonNull ).collect( Collectors.toSet());

But I think we can remove this storesByPkg cache in this single level reverse map solution because there is no full scan in it, which this storesByPkg cache is addressing to improve.